### PR TITLE
Refactor search to use react

### PR DESCRIPTION
**Pull request by mattheu**
Refactor existing search component to use React.

Not sure why I didn't write it like this to begin with. I think it was because I wasn't sure about using react to augment an existing component. However I think it's nice to keep the code in the theme a bit more consitent and its much neater now. 

Couple of queries.

* The live search is so fast my spinner is barely visible. Do people think its useful? I tried to debounce the removal to make the transition a little less frequent.
* The result items need the width of the container. This is to make the animation smoother. I've passed this in as a property, but doesn't really feel like a nice use of properties.

### CODE REVIEW

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import 'whatwg-fetch';
+
+var strings = {
+	label:       'Search',
+	button:      'Submit',
+	placeholder: 'Search the site…',
+};
+
+export default class SearchBar extends React.Component {
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: just a small personal preference, I like to `import React, { Component }` then just `extends Component`

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			query: this.props.query,
+			results: [],
+			focused: false,
+			loading: false,
+		};
+	}
+
+	render() {
+
+		var searchBarClasses = [ 'SearchBar' ];
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: could use `let` instead

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 1 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Will that affect my ability to use the React var? for stuff like `React.proptypes`

Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Oh you're importing both.

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+		return <form className={ searchBarClasses.join(' ') } role="search" method="get" onSubmit={ (e) => e.preventDefault() }>
+			<label className="SearchBar_Label" htmlFor="site-search">{ strings.label }</label>
+			<div className="SearchBar_Container">
+				<input ref="input" className="SearchBar_Field" type="search" id="site-search" placeholder={ strings.placeholder } value={ this.state.query } onChange={ () => this.onSearch() } onFocus={ () => this.onFocus() } onBlur={ () => this.onBlur() } />
+				<button className="SearchBar_Submit" onClick={ () => this.onSearch }>
+					{ strings.button }
+					<span className={ this.state.loading ? 'Loading Loading-Active' : 'Loading' }></span>
+				</button>
+				<SearchBarResults results={ this.state.results } containerWidth={ this.props.containerEl.offsetWidth } />
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: why are the search results inside the form?

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+		if ( this.state.loading ) {
+			searchBarClasses.push( 'SearchBar-Loading' );
+		}
+
+		return <form className={ searchBarClasses.join(' ') } role="search" method="get" onSubmit={ (e) => e.preventDefault() }>
+			<label className="SearchBar_Label" htmlFor="site-search">{ strings.label }</label>
+			<div className="SearchBar_Container">
+				<input ref="input" className="SearchBar_Field" type="search" id="site-search" placeholder={ strings.placeholder } value={ this.state.query } onChange={ () => this.onSearch() } onFocus={ () => this.onFocus() } onBlur={ () => this.onBlur() } />
+				<button className="SearchBar_Submit" onClick={ () => this.onSearch }>
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: I'd have this button be `type=submit` and do `<form ... onSubmit={this.onSearch}>`

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+					<span className={ this.state.loading ? 'Loading Loading-Active' : 'Loading' }></span>
+				</button>
+				<SearchBarResults results={ this.state.results } containerWidth={ this.props.containerEl.offsetWidth } />
+			</div>
+		</form>
+	}
+
+	onSearch( e ) {
+
+		var query = this.refs.input.value;
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: you should be able access `event.target.value` to get the search query from the input field so you wouldn't need the `ref` attribute

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+		this.setState( { query: query, loading: true } );
+		this.fetchResults( query )
+	}
+
+	fetchResults( query ) {
+		var api_endpoint = HMHandbookSearchSettings.api_endpoint;
+		var api_nonce    = HMHandbookSearchSettings.api_nonce;
+
+		api_endpoint += '?query=' + encodeURIComponent( query );
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: you can use `?query=${encodeURIComponent( query )}` https://developers.google.com/web/updates/2015/01/ES6-Template-Strings?hl=en

File: assets/src/scripts/components/SearchBar/SearchBarResult.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+
+/**
+ * History List.
+ */
+export default class SearchBar_Result extends React.Component {
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: I haven't seen underscores in class names in React, is this according to WP Standards?

File: assets/src/scripts/components/SearchBar/SearchBarResult.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+	render() {
+
+		var classes = [ 'SearchBar_Result' ];
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: could use `let`

File: searchform.php [hunk 9]
```diff
@@ -6,7 +6,6 @@
 		<div class="SearchBar_Container">
 			<input class="SearchBar_Field" type="search" id="site-search" placeholder="<?php esc_html_e( 'Search the site &hellip;', 'hm-handbook' ); ?>" value="<?php the_search_query(); ?>" name="s"/>
 			<button class="SearchBar_Submit"><?php esc_html_e( 'Submit', 'hm-handbook' ); ?></button>
```
Comment (pdewouters) [inline searchform.php]: not part of this PR but does the button need a `type="submit"`? or maybe just `submit_button('Submit', 'SearchBar_Submit')`

File: assets/src/scripts/theme.js [hunk 10]
```diff
@@ -18,11 +18,17 @@ if ( 'HMHandbookPageHistory' in window ) {
 
 }
 
+var searchBarContainers = document.querySelectorAll( '.StyleGuide_SearchContainer' );
 
-var searchBars = document.querySelectorAll( '.SearchBar' );
+for ( var i = 0; i < searchBarContainers.length; i++ ) {
```
Comment (pdewouters) [inline assets/src/scripts/theme.js]: `let i =0`

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 5 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Good call 👍

File: assets/src/scripts/components/SearchBar/SearchBarResult.js (hunk 7 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: I just named the class names after the CSS class names. If there are standards around this i'll change them.

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 4 above)
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: when there are lots of attributes I find this more legible
```
<input
  ref="input" 
  className="SearchBar_Field" 
  type="search"
  id="site-search" 
  placeholder={ strings.placeholder } 
  value={ this.state.query } onChange={ () => this.onSearch() } 
  onFocus={ () => this.onFocus() } 
  onBlur={ () => this.onBlur() } 
/>
```

File: assets/src/scripts/components/SearchBar/SearchBarResult.js (hunk 8 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: Is there any advantage to using `let` here? My understanding is it restricts the scope of the var to the current block, but since there are no other blocks in this function it makes no difference.

Again, if there are standards or other benefits to this, happy to change.

File: assets/src/scripts/theme.js (hunk 10 above)
Comment (mattheu) [inline assets/src/scripts/theme.js]: So here - totally see the benefit of let. Thanks :)

**Comment by pdewouters** [comment]
Just a few comments, might be worth considering this library for the class name logic https://github.com/JedWatson/classnames

I usually use `const` for declaring variables that won't get reassigned, and `let` in other cases

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 6 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Oh template strings are neat. Not used these before!

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 4 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: 👍

File: assets/src/scripts/components/SearchBar/SearchBarResult.js (hunk 8 above)
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: that's correct, I just interpreted these as prefer `let` over `var` https://gist.github.com/rmccue/8da274a14a03207849e4c1356255ed6a#variable-declaration , maybe I'm wrong

**Comment by mattheu** [comment]
Thanks for your review. Some good pointers on ES6. And I think the className library could be useful.

File: searchform.php (hunk 9 above)
Comment (mattheu) [inline searchform.php]: It does need` type="submit"`. But the `submit_button` function is only for use in the admin.

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 5 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Actually, now I've updated it to trigger `onSearch` on both change and form submit events, I think we do need the ref.

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 3 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Not sure... I'll play around with this to move it out.

File: assets/src/scripts/components/SearchBar/SearchBarResult.js (hunk 8 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: I suppose we can take the approach of making everything `let` by default and only giving it greater permission when needed. 

But I'm not so sure in this case. I'm concerned using `let` signals that it **shouldn't** be modified from outside the block of code in future. EG you shouldn't modify `classes` from within a callback or if statement? For this particular case, I wouldn't want to imply that shouldn't be done?

Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: Although in this case... i'm switching to that classNames helper and I think it makes sense to be `let` :)

File: searchform.php [hunk 11]
```diff
@@ -5,7 +5,7 @@
 
 		<div class="SearchBar_Container">
 			<input class="SearchBar_Field" type="search" id="site-search" placeholder="<?php esc_html_e( 'Search the site &hellip;', 'hm-handbook' ); ?>" value="<?php the_search_query(); ?>" name="s"/>
-			<?php submit_button( 'Submit', 'SearchBar_Submit' ) ?>
+			<button class="SearchBar_Submit"><?php esc_html_e( 'Submit', 'hm-handbook' ); ?></button>
```
Comment (pdewouters) [inline searchform.php]: oh, doesn't that work?

File: assets/src/scripts/components/SearchBar/SearchBar.js [hunk 12]
```diff
@@ -34,17 +34,17 @@ export default class SearchBar extends Component {
 			searchBarClasses.push( 'SearchBar-Loading' );
 		}
 
-		return <form className={ searchBarClasses.join(' ') } role="search" method="get" onSubmit={ (e) => e.preventDefault() }>
-			<label className="SearchBar_Label" htmlFor="site-search">{ strings.label }</label>
-			<div className="SearchBar_Container">
+		return <div className={ searchBarClasses.join(' ') }>
+			<form className="SearchBar_Container" role="search" method="get" onSubmit={ (e) => { e.preventDefualt(); this.onSearch() } }>
```
Comment (pdewouters) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: typo here `e.preventDefualt();`

File: searchform.php (hunk 11 above)
Comment (mattheu) [inline searchform.php]: Admin only!

File: assets/src/scripts/components/SearchBar/SearchBar.js (hunk 3 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBar.js]: Moved this out

File: assets/src/scripts/components/SearchBar/SearchBarResult.js (hunk 7 above)
Comment (mattheu) [inline assets/src/scripts/components/SearchBar/SearchBarResult.js]: Updated for consistency

**Comment by mattheu** [comment]
Think i've hit all of these things. I'm going to go ahead and merge!
